api/app/signals/apps/signals/v1/serializers.py

```diff
@@ -559,17 +559,22 @@ class Meta:
         extra_kwargs = {'file': {'write_only': True}}
 
     def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self._public = getattr(kwargs['context']['view'], 'is_public', False)
 
         # Set correct version of the links field (public/private)
-        if kwargs['context']['is_public']:
+        if self._public:
             self.serializer_url_field = PublicSignalAttachmentLinksField
         else:
             self.serializer_url_field = PrivateSignalAttachmentLinksField
 
-        super().__init__(*args, **kwargs)
-
     def create(self, validated_data):
-        signal = self.context['signal']
+        if self._public:
+            signal = self.context['view'].get_object()
+        else:
+            signal = self.context['view'].signal
```

Comment:
A bit difficult because of the choices that where made setting this up in the past. Will try to make this a better solution.

Reply:
I tried the suggested change (last thursday) and I the test suite did pass. You can pretty much copy the function body from the `signal` property.

Reply:
You forget that in this case the choice was made to overwrite the `get_queryset` of the private view (the public view still works how it should be) to return attachments instead of signals. Therefore the `get_object` will behave differently and will return any attachment it will find with the given id instead if the correct signal. So if we want this to work like how you suggest (which I mostly agree with) it will take a bit more work then just changing the `_signal` property to the `get_object` functionality.

However I will make the changes needed to get the same behaviour. I just wanted it to work first because it was deployed to acc broken and also not working locally.

Reply:
Done